lms/services/canvas_api/basic.py

```diff
@@ -0,0 +1,118 @@
+"""Low level access to the Canvas API."""
+
+from copy import deepcopy
+from urllib.parse import urlencode
+
+import requests
+from requests import RequestException, Session
+
+from lms.services import CanvasAPIError
+from lms.validation import ValidationError
+
+
+class BasicClient:
+    """
+    Provides low level access to the Canvas API.
+
+    Also supports:
+
+     * Pagination
+     * Apply schema to responses
+    """
+
+    PAGINATION_PER_PAGE = 1000
+    """The number of items to request at one time.
+
+     This only applies for calls which return more than one, and is subject
+     to a secret internal limit applied by Canvas (might be 100?)."""
+
+    PAGINATION_MAXIMUM_REQUESTS = 25
+    """The maximum number of calls to make before giving up."""
+
+    def __init__(self, canvas_host):
+        """
+        Create a new BasicClient for making calls to the Canvas API.
+
+        :param canvas_host: Hostname of the Canvas instance
+        """
+        self._session = Session()
+        self._canvas_host = canvas_host
+
+    def send(
+        # pylint: disable=too-many-arguments
+        self,
+        method,
+        path,
+        schema,
+        params=None,
+        headers=None,
+        url_stub="/api/v1",
+    ):
+        """
+        Make a request to the Canvas API and apply a schema to the response.
+
+        :param method: HTTP method to use
+        :param path: Path fragment to add to the end of the URL
+        :param schema: A schema object which this request must
+        :param params: URL parameters to include
+        :param headers: Headers to include
+        :param url_stub: Path prefix
+        :return: The result of applying the schema to the response
+        :raise CanvasAPIError: For any validation or request errors
```

Comment:
Grew another param here `url_stub`, but making this public avoided a lot of weird pain in the `AuthenticatedClient`.

Previously `path` and `params` were passed to `self.get_url()` which automatically included a `url_stub` for the API. The issue is the token paths don't use this so:

 * `AuthenticatedClient` had to call `get_url` itself
 * Then construct a new Request object itself
 * Then pass that to `send_and_validate`

If we just add this one parameter none of that is necessary, and the whole thing is simpler.